From 0f8928ff450f8255a36a08ed6f59468f982fa086 Mon Sep 17 00:00:00 2001
From: Konstantin Khlebnikov <khlebnikov@openvz.org>
Date: Tue, 1 Apr 2008 17:41:43 +0400
Subject: [PATCH 55/67] IPv6: make /proc/net/ipv6_route visible in VE

Show only rules of current VE (this logic already exists and works, entries
enumerated in private ve hash).

http://bugzilla.openvz.org/show_bug.cgi?id=857
---
 net/ipv6/route.c |   21 +++++++++++++++++++--
 1 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/net/ipv6/route.c b/net/ipv6/route.c
index 226ae36..74c434b 100644
--- a/net/ipv6/route.c
+++ b/net/ipv6/route.c
@@ -2476,6 +2476,23 @@ ctl_table ipv6_route_table[] = {
 
 #endif
 
+static int ip6_route_net_init(struct net *net)
+{
+	if (!proc_net_fops_create(net, "ipv6_route", 0, &ipv6_route_proc_fops))
+		return -ENOMEM;
+	return 0;
+}
+
+static void ip6_route_net_exit(struct net *net)
+{
+	proc_net_remove(net, "ipv6_route");
+}
+
+static struct pernet_operations ip6_route_net_ops = {
+	.init = ip6_route_net_init,
+	.exit = ip6_route_net_exit,
+};
+
 void __init ip6_route_init(void)
 {
 	ip6_dst_ops.kmem_cachep =
@@ -2484,7 +2501,6 @@ void __init ip6_route_init(void)
 	ip6_dst_blackhole_ops.kmem_cachep = ip6_dst_ops.kmem_cachep;
 
 	fib6_init();
-	proc_net_fops_create(&init_net, "ipv6_route", 0, &ipv6_route_proc_fops);
 	proc_net_fops_create(&init_net, "rt6_stats", S_IRUGO, &rt6_stats_seq_fops);
 #ifdef CONFIG_XFRM
 	xfrm6_init();
@@ -2496,15 +2512,16 @@ void __init ip6_route_init(void)
 	__rtnl_register(PF_INET6, RTM_NEWROUTE, inet6_rtm_newroute, NULL);
 	__rtnl_register(PF_INET6, RTM_DELROUTE, inet6_rtm_delroute, NULL);
 	__rtnl_register(PF_INET6, RTM_GETROUTE, inet6_rtm_getroute, NULL);
+	register_pernet_subsys(&ip6_route_net_ops);
 }
 
 void ip6_route_cleanup(void)
 {
+	unregister_pernet_subsys(&ip6_route_net_ops);
 #ifdef CONFIG_IPV6_MULTIPLE_TABLES
 	fib6_rules_cleanup();
 #endif
 #ifdef CONFIG_PROC_FS
-	proc_net_remove(&init_net, "ipv6_route");
 	proc_net_remove(&init_net, "rt6_stats");
 #endif
 #ifdef CONFIG_XFRM
-- 
1.5.4.3

